File: June/20-06-2024/Integral Points Inside Triangle.cs
//{ Driver Code Starts
// Initial Template for C#

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DriverCode {

class GFG {
    static void Main(string[] args) {
        int testcases; // Taking testcase as input
        testcases = Convert.ToInt32(Console.ReadLine());
        while (testcases-- > 0) // Looping through all testcases
        {

            long[] arr = new long[6];
            string elements = Console.ReadLine().Trim();
            elements = elements + " " + "0";
            arr = Array.ConvertAll(elements.Split(), long.Parse);
            long[] p = { arr[0], arr[1] };
            long[] q = { arr[2], arr[3] };
            long[] r = { arr[4], arr[5] };

            Solution obj = new Solution();
            long res = obj.InternalCount(p, q, r);
            Console.Write(res + "\n");
        }
    }
}
}
// } Driver Code Ends




// User function Template for C#

class Solution {
  //Your Code Start  
       public long gcd(long a, long b) {
        while (b != 0) {
            long temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    public long CountBoundaryPoints(long[] p, long[] q) {
        long x1 = p[0], y1 = p[1];
        long x2 = q[0], y2 = q[1];

        if (x1 == x2) {
            return Math.Abs(y2 - y1) - 1;
        }
        if (y1 == y2) {
            return Math.Abs(x2 - x1) - 1;
        }
        return gcd(Math.Abs(x2 - x1), Math.Abs(y2 - y1)) - 1;
    }

    public long InternalCount(long[] p, long[] q, long[] r) {
        long boundaryPoints = CountBoundaryPoints(p, q) +
                              CountBoundaryPoints(q, r) +
                              CountBoundaryPoints(r, p) + 3;

        long area = Math.Abs(p[0] * (q[1] - r[1]) +
                             q[0] * (r[1] - p[1]) +
                             r[0] * (p[1] - q[1]));

        long interiorPoints = (area - boundaryPoints + 2) / 2;

        return interiorPoints;
      //Your Code End
    }
}

File: June/20-06-2024/Integral Points Inside Triangle.py
#User function Template for python3

class Solution:
  #Your Code Start
    def gcd(self, a, b):
        while b:
            a, b = b, a % b
        return a

    def count_boundary_points(self, point1, point2):
        x1, y1 = point1
        x2, y2 = point2
        
        if x1 == x2:
            return abs(y2 - y1) - 1
        if y1 == y2:
            return abs(x2 - x1) - 1
        return self.gcd(abs(x2 - x1), abs(y2 - y1)) - 1

    def InternalCount(self, p, q, r):
        boundary_points = (self.count_boundary_points(p, q) +
                           self.count_boundary_points(p, r) +
                           self.count_boundary_points(q, r) + 3)
        
        area_twice = abs(p[0] * (q[1] - r[1]) +
                         q[0] * (r[1] - p[1]) +
                         r[0] * (p[1] - q[1]))
        
        interior_points = (area_twice - boundary_points + 2) // 2
        
        return interior_points

#Your Code End


#{ 
 # Driver Code Starts
#Initial Template for Python 3

import math
        
if __name__=='__main__':
    t=int(input())
    for _ in range(t):
        p=[0]*2
        q=[0]*2
        r=[0]*2
        p[0],p[1],q[0],q[1],r[0],r[1]=map(int,input().strip().split(" "))
        ob=Solution()
        ans=ob.InternalCount(p,q,r);
        print(ans)
# } Driver Code Ends
